### PR TITLE
fix(lru): correctly handle future cancellation

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As title. `lookup_dedup` will now spawn a task in the background, so that it won't deadlock.

Example: We have two futures running.

```
let future_1 = async move {
  lru.lookup("a", fetch_from_storage).await
}

let future_2 = async move {
  lru.lookup("a", fetch_from_storage).await
}
```

If we run them in the following way:

```
future_1.poll_next(); // so that we will cache miss, and LRU will call `fetch_from_storage` to get data.
future_2.poll_next(); // as request to "a" is in-flight, LRU will wait until future_1 completes the request.
drop(future_1);
loop {
  assert_eq!(future_2.poll_next(), Poll::pending); // future 2 will never return
}
```

Therefore, we spawn a tokio task in the background for every in-flight request, so that it won't be cancelled by users.

This PR also fixes some unit tests, so that they won't stuck in single-thread tokio runtime.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

close https://github.com/singularity-data/risingwave/issues/3909
close https://github.com/singularity-data/risingwave/pull/3907

### CODE REVIEW

**Comment by skyzh** [comment]
Need test before merging.

File: src/common/src/cache.rs [hunk 1]
```diff
@@ -766,25 +768,33 @@ impl<K: LruKey + Clone, T: LruValue> LruCache<K, T> {
     ) -> Result<Result<CachableEntry<K, T>, E>, RecvError>
     where
         F: FnOnce() -> VC,
```
Comment (BugenZhao) [inline src/common/src/cache.rs:770]: Seems async closure is unnecessary? We only need a future here. cc @Little-Wallace

Comment (skyzh) [reply src/common/src/cache.rs:770]: It's necessary. There's some variables that only need to be computed when cache miss. See other parts of the PR for more information.

Comment (skyzh) [reply src/common/src/cache.rs:770]: e.g. we need to call `self.meta_path` to compute the meta path ONLY when cache miss. But `&self` is not `'static`, so we can only compute it in this FnOnce closure and outside the `Future + 'static`.

**Comment by skyzh** [comment]
This would possibly fix TPC-H Q18.

**Review by TennyZhuang — APPROVED** [review]
Great work! Thanks!

**Comment by skyzh** [comment]
ref https://github.com/singularity-data/risingwave/issues/2580

**Comment by TennyZhuang** [comment]
Can you add some UTs for cancellation safety?

**Comment by skyzh** [comment]
> Can you add some UTs for cancellation safety?

You may add later. My test in https://github.com/singularity-data/risingwave/pull/3907 cannot work because now block cache requires tokio runtime.

File: src/common/src/cache.rs [hunk 2]
```diff
@@ -766,25 +768,33 @@ impl<K: LruKey + Clone, T: LruValue> LruCache<K, T> {
...
-                    let entry = self.insert(key, hash, charge, value);
-                    Ok(Ok(entry))
-                }
-                Err(e) => {
-                    self.clear_pending_request(&key, hash);
-                    Ok(Err(e))
-                }
-            },
+            LookupResult::Miss => {
+                let this = self.clone();
```
Comment (Little-Wallace) [inline src/common/src/cache.rs:781]: could we rewrite it as 
```rust
let ret = tokio::spawn(fetch_value).await;
```

Comment (skyzh) [reply src/common/src/cache.rs:781]: Nope. fetch_value itself is not 'static. The future it returns is 'static.

**Pull request merged**